### ISSUE TITLE
Absence of Surface Capabilities re-evaluation in Vulkan swapchain (re)building process.

### ISSUE THREAD

**Issue by NostraMagister** [ocornut/imgui#7842]
### Version/Branch of Dear ImGui:

ALL

### Back-ends:

imgui_impl_vulkan.cpp

### Compiler, OS:

all

### Full config/build information:

all

### Details:

The current swapchain (re)build code does not re-evaluate the physical device/surface capabilities before rebuilding the swapchain. This is absolutely needed in case users changed video driver settings such as frame rate, resolution, monitor swap, etc. At such point, Vulkan will, among others, update the minimum required images.

It should be done before the FrameSet is rebuild (in ALL branches).
Here is an example of how it could be done. It is important to remember that this is from my backend implementation that ONLY supports FIFO and MAILBOX present modes. Dear ImGui uses one more and might need some extra lines.
I left the comment for good understanding.

```
// Let capabilities set the this->vul_curSwapchainImages (and related vul_ImageCountChanged flag), this->vul_curWidth and this->vul_curHeight. 
// NOTE: If a windows was NOT minimized/hidden when we were called and it is now, then we may have a problem. See below.

this->Graphics_FindPhysicalDeviceSurfaceCapabilities(mla);
if (this->vul_curWidth == 0 || this->vul_curHeight == 0)
{
    // TODO: I have no solution yet for this. From the SDL3 supported platforms I think this 
    // should/could only happen on Android?
    printf("ATTENTION: Swapchain creation impossible while window width/height are (%d,%d)\n", this->vul_curWidth, this->vul_curHeight);
    return VK_ERROR_VALIDATION_FAILED_EXT;
}
```

```
// Find the CURRENT surface capabilities for a given physical device, useful for Graphical applications only. This is called from 
// SelectPhysicalDevice() to initialize, then it is called by code that needs an update of the capabilities like the CreateSwapchain() method. 
// The method assumes that the presentation mode, and hence vul_min/maxSwapchainImages, the require/allowed images in the swapchain for that mode, 
// has been set already (which is the case because SelectPhysicalDevice() does that). 
// The method calculates and adjusts the following things as needed:
// - the CURRENT number of images to use for swapchain creation (based on 'wanted' and 'possible', and stores it in vul_curSwapchainImages. 
// - the CURRENT image size (width/height) to use for swapchain creation, and stores it in vul_curWidth and vul_curHeight.
// If the current size is unlimited (0xFFFFFFFF) the MLA (SDL) GetWindowSizeInPixels() method is used to find the current windows size.
// 
// VK_SUCCESS if all values are based on 'wanted', not delimited by 'possible'bounds. 
// VK_NOT_READY will be returned if this->vul_Surface or this->vul_PhysicalDevice is a VK_NULL_HANDLE on entry. Protects against logic errors.
// VK_ERROR_IMAGE_USAGE_NOT_SUPPORTED_KHR this is VK_SUCCESS but values where delimited by 'possible' bounds and deviate from 'wanted'.
VkResult MetaGAPI::Graphics_FindPhysicalDeviceSurfaceCapabilities(MetaMLA* mla)
{
    VkResult err = VK_SUCCESS;
    if (this->vul_Surface == VK_NULL_HANDLE) return VK_NOT_READY;
    if (this->vul_PhysicalDevice == VK_NULL_HANDLE) return VK_NOT_READY;

    // The vkGetPhysicalDeviceSurfaceCapabilitiesKHR function return what the CURRENT 'possible' capabilities are. These can change due 
    // to user actions such as changing display resolution, maximizing window, changing some driver parameters, etc. E.g., increasing 
    // the drivers frame rate may increase the minimum number of images the swapchain requires, depending on presentation mode.
    // When for instance the swapchain is build (see this->CreateSwapchain()) we will need to get the capabilities at THAT moment.
    // For the explanation of the different members of the structure see 
    // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSurfaceCapabilitiesKHR.html
    VkSurfaceCapabilitiesKHR surfaceCapabilities;
    vkGetPhysicalDeviceSurfaceCapabilitiesKHR(this->vul_PhysicalDevice, this->vul_Surface, &surfaceCapabilities);

    // ********** IMAGE WIDTH/HEIGHT **********
    
    // The surfaceCapabilities.currentExtent returns what the surface size (width/height) is that we need to use based on current Media Layer window size. 
    // If currentExtend.width or .height is 0xFFFFFFFF or std::numeric_limits<uint32_t>::max(), then we may choose the width and height as long 
    // as it is within bounds of the provided min/max, also in available through VkSurfaceCapabilitiesKHR. 
    // ATTENTION: See also https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSwapchainPresentScalingCreateInfoEXT.html
    //            width/height might be (0,0) on some platforms (e.g. when window minimizes) which would be a problem during swapchain building. 
    //            See Graphics_CreateSwapchain() for how we handle that to make it platform safe.
    
    // Check if we must use the exact size that is in surfaceCapabilities.currentExtent or if we may chose ourselves (value 0xFFFFFFFF).
    if (surfaceCapabilities.currentExtent.width == std::numeric_limits<uint32_t>::max()) 
    {   // We can choose the width/height. let's get them from the MLA (SDL) and bring them within bounds.
        int W, H;
        mla->GetWindowSizeInPixels(&W, &H); // former SDL2 SDL_Vulkan_GetDrawableSize() equivalent.
        this->vul_curWidth = std::max(surfaceCapabilities.minImageExtent.width, std::min(surfaceCapabilities.maxImageExtent.width, static_cast<uint32_t>(W)));
        this->vul_curHeight = std::max(surfaceCapabilities.minImageExtent.height, std::min(surfaceCapabilities.maxImageExtent.height, static_cast<uint32_t>(H)));
    }
    else 
    {   // We must use the exact content off currentExtent.
        this->vul_curWidth = surfaceCapabilities.currentExtent.width;
        this->vul_curHeight = surfaceCapabilities.currentExtent.height;
    }
        
    // ********** IMAGES IN SWAPCHAIN **********

    // This implementation defaults to minimum 'wanted' number of images based on the chosen presentation mode (see Graphics_SelectPresentationMode()). 
    // It allows the user to set a maximum 'wanted' number of images (See SetMaximumSwapchainImages()) and if not set, defaults it to the minimum.
    // The code below calculates the CURRENT maximum 'possible' 'wanted' values. The function ALWAYS succeeds but the return value will reflect whether we 
    // got what we wanted VK_SUCCESS or whether we got what was possible VK_ERROR_IMAGE_USAGE_NOT_SUPPORTED_KHR. The implementation can use this to know 
    // if, and how, it can continue processing. It is not expected that if MAILBOX presentation mode is available on the Vulkan Instance (using the 
    // needed extension) that VkSurfaceCapabilitiesKHR would return a minImageCount less then 3, however, since drivers sometimes behave strange
    // we are prepared for it.
    
    // Calculate the minimal wanted and possible number of swapchain images. 'possible' capabilities precede on 'wanted'. Will be minimum 1.
    std::uint32_t min_swi = this->vul_minSwapchainImages > surfaceCapabilities.minImageCount ? this->vul_minSwapchainImages : surfaceCapabilities.minImageCount;
    
    // Calculate the maximal wanted and possible number of swapchain images. 'possible' capabilities precede on 'wanted'. Unlimited capabilities becomes wanted maximum.
    // A value of 0 means that there is no limit on the number of images, though there may be limits related to the total amount of memory used by presentable images.
    std::uint32_t max_swi = surfaceCapabilities.maxImageCount != 0 ? surfaceCapabilities.maxImageCount : this->vul_maxSwapchainImages;
    
    // In case surfaceCapabilities.maxImageCount was not 0 but it is higher then this->vul_maxSwapchainImages, it must be topped off.
    if (max_swi > this->vul_maxSwapchainImages) max_swi = this->vul_maxSwapchainImages;
    
    // In case at this point the minimum is higher then the calculated maximum, the minimum is reduced to the maximum.
    // We add this to make it watertide with different drivers. cap 2:2 wanted 3:5 would be such case that at this point would have 3:2.
    //if (min_swi > max_swi) min_swi = max_swi; // No longer needed, see next line, we use max_swi.

    // Set the number of images 'to be'/'that can be' used for the swapchain build.
    this->vul_curSwapchainImages = max_swi;

    // Set the return value
    if (this->vul_curSwapchainImages < this->vul_minSwapchainImages) err = VK_ERROR_IMAGE_USAGE_NOT_SUPPORTED_KHR;
    
    // Debug Only
    printf("Vulkan Surface Size Capabilities : curWidth %d\tcurHeight %d\n", this->vul_curWidth, this->vul_curHeight);
    printf("Vulkan Surface Image Capabilities: minSwapchainImages %d\tmaxSwapchainImages %d\tcurSwapchainImages %d\n", this->vul_minSwapchainImages, this->vul_maxSwapchainImages, this->vul_curSwapchainImages);

    return err;
}
```

**Comment by ocornut** [comment]
I appreciate you opening 5+ new vulkan related issues and commenting on others. I really appreciate the detailed messages.
But if I have to be pragmatic/realistic, in spite of upmost goodwill the most likely scenario that those messages are going stay around for a very long time unhandled. Backends have been attention hog and Vulkan in particular to the point of being problematic. It ends up being too overwhelming to deal with.

I think the best you could do would be to 1) finish your review to have a larger view of everything. 2) make PRs as nicely thoughtful as your existing posts.

Bonus, I would also appreciate if for every requests you could specify if are coming from an angle of having stumbled on a real blocking issue, or if it stems from really docs/specs.

**Comment by NostraMagister** [comment]
OK, i will do it that way. 

The tickets are contributions because I do not depend on any fixes because I will not have the Dear ImGui courtesy backend code in use. Furthermore, i try to be careful with what i propose because you maintain many backends and i only look at all this from an SDL3/Vulkan perspective. I started writing the tickets because I found comments like:

```
// Select Present Mode
// FIXME-VULKAN: Even thought mailbox seems to get us maximum framerate with a single window, it halves framerate with a second window etc. (w/ Nvidia and SDK 1.82.1)
```

Its kind of in my nature wanting to find the 'why' :)

**Comment by ocornut** [comment]
It’s going to be quite difficult for us to look into suggestion if the user is not using the backend. I hope you’ll want to consider using the backend once you understand the issue you had.

**Comment by NostraMagister** [comment]
Yes, i'll share everything i find but i need to be sure and not maybe sure otherwise its a loss of time for everyone.

**Comment by HackettJP** [comment]
@NostraMagister ,I like to put my hand question ? I could be totally wrong which is good :)

I'm trying to unwind *Learn* without my head exploding :)  (Imgui Vulkan.cpp)<-->(Vulkan Tutorials.cpp) and creating a hybrid version to deep understand everything, reading all the issues requests
 
I have  test app with 2d images then working out adding compute and 3d rendering.

Seems to me probably totally wrong as vulkan is build for one scenario,different scenario 'yep rewrite'.

it is if you need a APP->imgui->*imgui_vulkan_backend->*intermidate_new->vulkan, *insert deep breath lots of (coffee,tea),  session, abstraction layer, configuration.

With the imgui_vulkan_backend more of a configuration, parameters dependent on scenario, than actual implemention of the command, swapchains, fences, which vulkan so loves, reminds me so much of async ring buffers and PCI device drivers in SUN Solaris OS.

Vulkan requires so much understanding upfront 'you have already built it before you built it' to know of the finer details that can hurt you later ?

I don't mind chatting off-line or disappearing :), or help out if need be.

J.

**Comment by NostraMagister** [comment]
@HackettJP, not sure if I understand exactly what you are saying or asking. 

This thread is about the fact that Vulkan specifications require VkSurfaceCapabilitiesKHR to be called to make sure that the swapchain is correctly  rebuild. If a user changes video driver parameters while a Dear ImGui based application using the courtesy Vulkan backend is running, it may change width, height and minimum no. required images. Hence, one cannot just rebuild the swapchain based on the number of images required by the present mode. Set your screen in the highest resolution and run your application with a larger window size (width/height), then let the application running and change the video drivers setting to your lowest resolution, and maybe you will crash because the capabilities width/height changed.

If your question was about writing your own backend, then I advice to follow the suggestions of the Dear ImGui team and use one of their courtesy backends and start from there. If you do that you will have reached your goal quoted below. 

> creating a hybrid version to deep understand everything, 

As the Dear ImGui in-code comments suggest, this is not a simple task, buy more coffee :)
I hope this helps.
